File: Java/Combinations.java
public class Solution {
    /**
     * @param n: Given the range of numbers
     * @param k: Given the numbers of combinations
     * @return: All the combinations of k numbers out of 1..n
     */
    public void helper(List<List<Integer>> res, List<Integer> cur, int num, int pos, int end) {
        if (num == 0) {
            res.add(new ArrayList<Integer>(cur));
            return;
        }
        for (int i = pos; i <= end; i ++){
            cur.add(i);
            helper(res, cur, num - 1, i + 1, end);
            cur.remove(cur.size() - 1);
        }
    }
    public List<List<Integer>> combine(int n, int k) {
        // write your code here
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        List<Integer> cur = new ArrayList<Integer>();
        if (k > n || n <= 0) return res;
        helper(res, cur, k, 1, n);
        return res;
    }
}

File: Python/Combinations.py
class Solution:
    """
    @param n: Given the range of numbers
    @param k: Given the numbers of combinations
    @return: All the combinations of k numbers out of 1..n
    """
    def helper(self, n, pos, k, res, cur):
        if k == 0:
            res.append(cur[:])
            return
        if k < 0 or pos > n:
            return
        for i in range(pos, n + 1):
            cur.append(i)
            self.helper(n, i + 1, k - 1, res, cur)
            cur.pop()

    def combine(self, n, k):
        # write your code here
        res = []
        cur = []
        self.helper(n, 1, k, res, cur)
        return res